src/server/journal/journal_slice.cc

```diff
@@ -190,28 +190,25 @@ void JournalSlice::AddLogRecord(const Entry& entry, bool await) {
   {
     std::shared_lock lk(cb_mu_);
     DVLOG(2) << "AddLogRecord: run callbacks for " << entry.ToString()
-             << " num callbacks: " << change_cb_arr_.size();
+             << " num callbacks: " << change_cbs_.size();
 
-    for (const auto& k_v : change_cb_arr_) {
+    for (const auto& k_v : change_cbs_) {
       k_v.second(*item, await);
     }
   }
 }
 
 uint32_t JournalSlice::RegisterOnChange(ChangeCallback cb) {
-  // mutex lock isn't needed due to iterators are not invalidated
-  uint32_t id = next_cb_id_++;
-  change_cb_arr_.emplace_back(id, std::move(cb));
-  return id;
+  lock_guard lk(cb_mu_);
```

Comment:
Why do you need it. I thought with global transaction we don't need it anymore

Reply:
True, but we still need the mutex for deletes, becase we unregister arbitrarily

Reply:
But for db_slice we don't need it?

Reply:
Yes, because there is no code path form those callbacks that can unregister itself

Reply:
Somewhat fragile, I agree